core/trino-main/src/main/java/io/trino/operator/DeduplicationExchangeClientBuffer.java

```diff
@@ -0,0 +1,348 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator;
+
+import com.google.common.base.VerifyException;
+import com.google.common.collect.ArrayListMultimap;
+import com.google.common.collect.ListMultimap;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.SettableFuture;
+import io.airlift.units.DataSize;
+import io.trino.execution.TaskId;
+import io.trino.execution.buffer.SerializedPage;
+import io.trino.spi.TrinoException;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.Executor;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Throwables.throwIfUnchecked;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.util.concurrent.Futures.nonCancellationPropagating;
+import static io.trino.operator.RetryPolicy.QUERY;
+import static io.trino.operator.RetryPolicy.TASK;
+import static io.trino.spi.StandardErrorCode.NOT_SUPPORTED;
+import static java.lang.Math.max;
+import static java.util.Objects.requireNonNull;
+
+public class DeduplicationExchangeClientBuffer
+        implements ExchangeClientBuffer
+{
+    private final Executor executor;
+    private final long bufferCapacityInBytes;
+    private final RetryPolicy retryPolicy;
+
+    private final SettableFuture<Void> blocked = SettableFuture.create();
+    @GuardedBy("this")
+    private final Set<TaskId> allTasks = new HashSet<>();
+    @GuardedBy("this")
+    private boolean noMoreTasks;
+    @GuardedBy("this")
+    private final Set<TaskId> finishedTasks = new HashSet<>();
```

Comment:
maybe rename to `successfulTasks`. Otherwise it is not obvious if it should not also contain failed tasks.